src/client/Client.cc

```diff
@@ -9823,18 +9823,180 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
+    in->mark_caps_dirty(CEPH_CAP_FILE_WR);
+    clnt->check_caps(in, 0);
+  }
+
+  onuninlinefinished = true;
+  onuninlinefinished_r = r;
+  fini = try_complete();
+
+  clnt->client_lock.unlock();
+
+  if (fini)
+    delete this;
+}
+
+bool Client::C_Read_Finisher::try_complete()
+{
+  if (onuninlinefinished && iofinished) {
+    if (have) {
+      clnt->put_cap_ref(in, CEPH_CAP_FILE_RD);
+    }
+    if (movepos) {
+      clnt->unlock_fh_pos(f);
+    }
+
+     if (onuninlinefinished_r >= 0 || onuninlinefinished_r == -CEPHFS_ECANCELED)
+      onfinish->complete(iofinished_r);
+    else
+      onfinish->complete(onuninlinefinished_r);
+    return true;
+  }
+  return false;
+}
+
+void Client::C_Read_Sync_Async::retry()
+{
+  filer->read_trunc(in->ino, &in->layout, in->snapid, pos, left, &tbl, 0,
+                    in->truncate_size, in->truncate_seq, this);
+}
+
+void Client::C_Read_Sync_Async::finish(int r)
+{
+  clnt->client_lock.lock();
+
+  if (r == -CEPHFS_ENOENT) {
+    // if we get ENOENT from OSD, assume 0 bytes returned
+    goto success;
+  } else if (r < 0) {
+    // pass error to caller
+    goto error;
+  }
+
+  if (tbl.length()) {
+    r = tbl.length();
+
+    read += r;
+    pos += r;
+    left -= r;
+    bl->claim_append(tbl);
+  }
+
+  // short read?
+  if (r >= 0 && r < wanted) {
+    if (pos < in->size) {
+      // zero up to known EOF
+      int64_t some = in->size - pos;
+      if (some > left)
+        some = left;
+      auto z = buffer::ptr_node::create(some);
+      z->zero();
+      bl->push_back(std::move(z));
+      read += some;
+      pos += some;
+      left -= some;
+      if (left == 0)
+        goto success;
+    }
+
+    clnt->put_cap_ref(in, CEPH_CAP_FILE_RD);
+    // reverify size
+    {
+      r = clnt->_getattr(in, CEPH_STAT_CAP_SIZE, f->actor_perms);
+      if (r < 0)
+        goto error;
+    }
+
+    // eof?  short read.
+    if ((uint64_t)pos >= in->size)
+      goto success;
+
+    {
+      int have2 = 0;
+      r = clnt->get_caps(f, CEPH_CAP_FILE_RD, have, &have2, -1);
+      if (r < 0) {
+        goto error;
+      }
+    }
+
+    wanted = left;
+    retry();
+    clnt->client_lock.unlock();
+    return;
+  }
+
+success:
+
+  r = read;
+
+error:
+
+  onfinish->complete(r);
+  fini = true;
+
+  clnt->client_lock.unlock();
+}
+
+int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl,
+                      Context *onfinish)
 {
   ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 
   int want, have = 0;
   bool movepos = false;
-  std::unique_ptr<C_SaferCond> onuninline;
+  std::unique_ptr<Context> onuninline;
+  std::unique_ptr<Context> iofinish = nullptr;
+  std::unique_ptr<C_Read_Finisher> crf = nullptr;
   int64_t rc = 0;
   const auto& conf = cct->_conf;
   Inode *in = f->inode.get();
   utime_t lat;
   utime_t start = ceph_clock_now(); 
+  CRF_onuninline *crf_onuninline = NULL;
+  CRF_iofinish *crf_iofinish = NULL;
+  C_SaferCond *cond_onuninline = NULL;
```

Comment:
`s/NULL/nullptr/`

would be idiomatic c++

Reply:
fixed